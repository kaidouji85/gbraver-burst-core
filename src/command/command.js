// @flow
import type {BatteryCommand} from "./battery";
import type {BurstCommand} from "./burst";
import type {EmptyCommand} from "./empty-command";
import type {PlayerId} from "..";
import type {PilotSkillCommand} from "./pilot-skill";

/** コマンド */
export type Command = EmptyCommand | BatteryCommand | BurstCommand | PilotSkillCommand;

/**
 * クイックコマンド
 */
export type QuickCommand = BurstCommand | PilotSkillCommand;

/**
 * 指定したコマンドをクイックコマンドにキャストする
 * キャストできない場合はnullを返す
 *
 * @param command キャストするコマンド
 * @return キャスト結果
 */
export function castQuickCommand(command: Command): ?QuickCommand {
  switch (command.type) {
    case 'BURST_COMMAND':
    case 'PILOT_SKILL_COMMAND':
      return command;
    default:
      return null;
  }
}

/**
 * プレイヤーコマンド
 *
 * @typeparam {X} コマンド
 */
export type PlayerCommandX<X> = {
  /** プレイヤーID */
  playerId: PlayerId,
  /** コマンド */
  command: X
};

/** プレイヤーコマンド */
export type PlayerCommand = PlayerCommandX<Command>;