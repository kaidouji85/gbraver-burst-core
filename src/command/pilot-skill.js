// @flow

/**
 * パイロットスキルコマンド
 */
export type PilotSkillCommand = {
  type: 'PilotSkillCommand'
}