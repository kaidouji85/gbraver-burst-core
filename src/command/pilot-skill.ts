/**
 * パイロットスキルコマンド
 */
export type PilotSkillCommand = {
  type: "PILOT_SKILL_COMMAND";
};