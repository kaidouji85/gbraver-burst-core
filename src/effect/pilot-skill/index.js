// @flow

import type {GameState, GameStateX, PilotSkill, PilotSkillEffect, PilotSkillEffectX, PlayerId} from "../..";
import type {BuffPowerSkill, RecoverBatterySkill} from "../../player/pilot";
import {recoverBattery} from "./recover-battery";
import {buffPower} from "./buff-power";

/**
 * パイロットスキルを発動する
 *
 * @param lastState 最新の状態
 * @param invokerId パイロットスキルを発動するプレイヤー
 * @return 更新結果、実行不可能な場合はnullを返す
 */
export function pilotSkill(lastState: GameState, invokerId: PlayerId): ?GameStateX<PilotSkillEffect> {
  const donePilotSkill = pilotSkillEffect(lastState, invokerId);
  if (!donePilotSkill) {
    return null;
  }
  return disablePilotSkill(donePilotSkill);
}

/**
 * パイロットスキルを適用する
 *
 * @param lastState 最新の状態
 * @param invokerId パイロットスキルを発動するプレイヤー
 * @return 更新結果、実行不可能な場合はnullを返す
 */
function pilotSkillEffect(lastState: GameState, invokerId: PlayerId): ?GameStateX<PilotSkillEffect> {
  const invoker = lastState.players.find(v => v.playerId === invokerId);
  if (!invoker) {
    return null;
  }

  if (invoker.pilot.skill.type === 'RecoverBatterySkill') {
    const castedSkill: RecoverBatterySkill = invoker.pilot.skill;
    const updated =  recoverBattery(lastState, invokerId, castedSkill);
    return updated ? upcastPilotSkillEffect(updated) : null;
  }

  if (invoker.pilot.skill.type === 'BuffPowerSkill') {
    const castedSkill: BuffPowerSkill = invoker.pilot.skill;
    const updated = buffPower(lastState, invokerId, castedSkill);
    return updated ? upcastPilotSkillEffect(updated) : null;
  }

  return null;
}

/**
 * パイロットスキルを使用済みにする
 *
 * @param lastState 最新状態
 * @return 更新結果、実行不可能な場合はnullを返す
 */
function disablePilotSkill(lastState: GameStateX<PilotSkillEffect>): ?GameStateX<PilotSkillEffect> {
  const invoker = lastState.players.find(v => v.playerId === lastState.effect.invokerId);
  if (!invoker) {
    return null;
  }

  const updatedInvoker = {
    ...invoker,
    pilot: {
      ...invoker.pilot,
      enableSkill: false
    }
  };
  const updatedPlayers = lastState.players
    .map(v => v.playerId === updatedInvoker.playerId ? updatedInvoker : v);
  return {
    ...lastState,
    players: updatedPlayers
  };
}

/**
 * パイロットスキル発動ステートにアップキャストする
 *
 * @param origin キャスト前
 * @return キャスト結果
 */
function upcastPilotSkillEffect<X: PilotSkill>(origin: GameStateX<PilotSkillEffectX<X>>): GameStateX<PilotSkillEffect> {
  return ((origin: any): GameStateX<PilotSkillEffectX<PilotSkill | typeof origin.effect.skill>>);
}
