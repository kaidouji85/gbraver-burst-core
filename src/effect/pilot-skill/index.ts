import type { PlayerId } from "../../player/player";
import type { GameState, GameStateX } from "../../state/game-state";
import { PlayerState } from "../../state/player-state";
import { batteryBoost } from "./battery-boost";
import { batteryEnchantment } from "./battery-enchantment";
import { buffPower } from "./buff-power";
import { damageHalvedSkill } from "./damage-halved-skill";
import type { PilotSkillEffect } from "./pilot-skill-effect";
import { recoverBattery } from "./recover-battery";

/**
 * パイロットスキルを適用する
 * @param lastState 最新の状態
 * @param invokerId パイロットスキルを発動するプレイヤー
 * @return 更新結果、実行不可能な場合は例外を返す
 */
function pilotSkillEffect(
  lastState: GameState,
  invokerId: PlayerId,
): GameStateX<PilotSkillEffect> {
  const invoker = lastState.players.find((v) => v.playerId === invokerId);
  if (!invoker) {
    throw new Error("not found pilot skill invoker");
  }

  if (invoker.pilot.skill.type === "RecoverBatterySkill") {
    return recoverBattery(lastState, invokerId, invoker.pilot.skill);
  }

  if (invoker.pilot.skill.type === "BuffPowerSkill") {
    return buffPower(lastState, invokerId, invoker.pilot.skill);
  }

  if (invoker.pilot.skill.type === "BatteryEnchantmentSkill") {
    return batteryEnchantment(lastState, invokerId, invoker.pilot.skill);
  }

  if (invoker.pilot.skill.type === "DamageHalvedSkill") {
    return damageHalvedSkill(lastState, invokerId, invoker.pilot.skill);
  }

  if (invoker.pilot.skill.type === "BatteryBoostSkill") {
    return batteryBoost(lastState, invokerId, invoker.pilot.skill);
  }

  throw new Error("not found pilot skill");
}

/**
 * パイロットスキルを使用済みにする
 * @param lastState 最新状態
 * @return 更新結果
 */
function disablePilotSkill(
  lastState: GameStateX<PilotSkillEffect>,
): GameStateX<PilotSkillEffect> {
  const updateInvoker = (invoker: PlayerState) => ({
    ...invoker,
    pilot: { ...invoker.pilot, enableSkill: false },
  });
  const updatedPlayers = lastState.players.map((v) =>
    v.playerId === lastState.effect.invokerId ? updateInvoker(v) : v,
  );
  return { ...lastState, players: updatedPlayers };
}

/**
 * パイロットスキルを発動する
 * @param lastState 最新の状態
 * @param invokerId パイロットスキルを発動するプレイヤー
 * @return 更新結果、実行不可能な場合は例外を返す
 */
export function pilotSkill(
  lastState: GameState,
  invokerId: PlayerId,
): GameStateX<PilotSkillEffect> {
  const donePilotSkill = pilotSkillEffect(lastState, invokerId);
  return disablePilotSkill(donePilotSkill);
}
