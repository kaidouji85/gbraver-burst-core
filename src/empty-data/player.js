// @flow

import type {PlayerState} from "../state/player-state";
import {EMPTY_ARMDOZER_STATE} from "./armdozer";
import {EMPTY_PILOT_STATE} from "./pilot";

/** 空のプレイヤーステート */
export const EMPTY_PLAYER_STATE: PlayerState = {
  playerId: '',
  armdozer: EMPTY_ARMDOZER_STATE,
  pilot: EMPTY_PILOT_STATE,
};
