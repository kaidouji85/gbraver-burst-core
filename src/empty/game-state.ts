import type { GameState } from "../state/game-state";

/** 空のゲームステート */
export const EMPTY_GAME_STATE: GameState = {
  players: [],
  activePlayerId: "",
  effect: {
    name: "StartGame"
  }
};