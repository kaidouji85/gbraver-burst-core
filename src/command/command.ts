import { z } from "zod";

import type { BatteryCommand } from "./battery";
import { BatteryCommandSchema } from "./battery";
import type { BurstCommand } from "./burst";
import { BurstCommandSchema } from "./burst";
import type { EmptyCommand } from "./empty-command";
import { EmptyCommandSchema } from "./empty-command";
import type { PilotSkillCommand } from "./pilot-skill";
import { PilotSkillCommandSchema } from "./pilot-skill";

/** コマンド */
export type Command =
  | EmptyCommand
  | BatteryCommand
  | BurstCommand
  | PilotSkillCommand;

/** コマンド zodスキーマ */
export const CommandSchema = z.union([
  EmptyCommandSchema,
  BatteryCommandSchema,
  BurstCommandSchema,
  PilotSkillCommandSchema,
]);

/**
 * コマンドにパースする
 * @param origin パース元
 * @return パース結果、失敗した場合はnull
 */
export function parseCommand(origin: unknown): Command | null {
  const result = CommandSchema.safeParse(origin);
  return result.success ? result.data : null;
}
