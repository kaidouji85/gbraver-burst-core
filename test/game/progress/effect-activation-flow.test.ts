import {
  BatteryCommand,
  Burst,
  BurstCommand,
  ForceTurnEnd,
  GameState,
  PlayerCommand,
  PlayerId,
  PlayerState,
} from "../../../src";
import { PilotSkillCommand } from "../../../src/command/pilot-skill";
import { EMPTY_GAME_STATE } from "../../../src/empty/game-state";
import { EMPTY_PLAYER_STATE } from "../../../src/empty/player";
import { effectActivationFlow } from "../../../src/game/progress/effect-activation-flow";

/**
 * テストプレイヤーを生成うる
 * @param playerId プレイヤーID
 * @param burst バースト
 */
const createPlayer = (
  playerId: PlayerId,
  burst: Burst = {
    type: "RecoverBattery",
    recoverBattery: 5,
    turnStartBatteryCorrect: 1,
  },
): PlayerState => ({
  ...EMPTY_PLAYER_STATE,
  playerId,
  armdozer: {
    ...EMPTY_PLAYER_STATE.armdozer,
    burst,
  },
});

/** 強制ターン終了 */
const FORCE_TURN_END: ForceTurnEnd = {
  type: "ForceTurnEnd",
  recoverBattery: 2,
};

/** バッテリーコマンド */
const BATTERY_COMMAND: BatteryCommand = {
  type: "BATTERY_COMMAND",
  battery: 1,
};

/** バーストコマンド */
const BURST_COMMAND: BurstCommand = {
  type: "BURST_COMMAND",
};

/** パイロットスキルコマンド */
const PILOT_SKILL_COMMAND: PilotSkillCommand = {
  type: "PILOT_SKILL_COMMAND",
};

test("攻撃側だけが効果発動した場合、防御側は何もせずにフローが進む", () => {
  const attacker = createPlayer("attacker");
  const defender = createPlayer("defender");
  const state: GameState = {
    ...EMPTY_GAME_STATE,
    players: [defender, attacker],
    activePlayerId: attacker.playerId,
  };
  const commands: [PlayerCommand, PlayerCommand] = [
    { playerId: attacker.playerId, command: BURST_COMMAND },
    { playerId: defender.playerId, command: BATTERY_COMMAND },
  ];
  const result = effectActivationFlow(state, commands);
  expect(result).toMatchSnapshot();
});

test("防御側だけが効果発動した場合、攻撃側はなにもせずにフローが進む", () => {
  const attacker = createPlayer("attacker");
  const defender = createPlayer("defender");
  const state: GameState = {
    ...EMPTY_GAME_STATE,
    players: [defender, attacker],
    activePlayerId: attacker.playerId,
  };
  const commands: [PlayerCommand, PlayerCommand] = [
    { playerId: attacker.playerId, command: BATTERY_COMMAND },
    { playerId: defender.playerId, command: PILOT_SKILL_COMMAND },
  ];
  const result = effectActivationFlow(state, commands);
  expect(result).toMatchSnapshot();
});

test("攻撃側、防御側の両方で効果を発動した場合、両方の効果を発動してフローが進む", () => {
  const attacker = createPlayer("attacker");
  const defender = createPlayer("defender");
  const state: GameState = {
    ...EMPTY_GAME_STATE,
    players: [defender, attacker],
    activePlayerId: attacker.playerId,
  };
  const commands: [PlayerCommand, PlayerCommand] = [
    { playerId: attacker.playerId, command: BURST_COMMAND },
    { playerId: defender.playerId, command: PILOT_SKILL_COMMAND },
  ];
  const result = effectActivationFlow(state, commands);
  expect(result).toMatchSnapshot();
});

test("攻撃側が強制ターン終了を発動した場合、防御側の効果は発動されない", () => {
  const attacker = createPlayer("attacker", FORCE_TURN_END);
  const defender = createPlayer("defender");
  const state: GameState = {
    ...EMPTY_GAME_STATE,
    players: [defender, attacker],
    activePlayerId: attacker.playerId,
  };
  const commands: [PlayerCommand, PlayerCommand] = [
    { playerId: attacker.playerId, command: BURST_COMMAND },
    { playerId: defender.playerId, command: PILOT_SKILL_COMMAND },
  ];
  const result = effectActivationFlow(state, commands);
  expect(result).toMatchSnapshot();
});

test("防御側が強制ターン終了を発動した場合、攻撃側の効果は発動されるが攻撃プレイヤーが交換される", () => {
  const attacker = createPlayer("attacker");
  const defender = createPlayer("defender", FORCE_TURN_END);
  const state: GameState = {
    ...EMPTY_GAME_STATE,
    players: [defender, attacker],
    activePlayerId: attacker.playerId,
  };
  const commands: [PlayerCommand, PlayerCommand] = [
    { playerId: attacker.playerId, command: BURST_COMMAND },
    { playerId: defender.playerId, command: BURST_COMMAND },
  ];
  const result = effectActivationFlow(state, commands);
  expect(result).toMatchSnapshot();
});