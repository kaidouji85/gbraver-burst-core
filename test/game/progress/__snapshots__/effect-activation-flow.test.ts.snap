// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`一人だけ効果適用する場合でも正しく処理される: one-player-effective 1`] = `
[
  {
    "activePlayerId": "attacker",
    "effect": {
      "burst": {
        "recoverBattery": 5,
        "turnStartBatteryCorrect": 1,
        "type": "RecoverBattery",
      },
      "burstPlayer": "attacker",
      "name": "BurstEffect",
    },
    "players": [
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "defender",
      },
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [
            {
              "correctBattery": 1,
              "period": {
                "type": "SpecialPeriod",
              },
              "type": "TurnStartBatteryCorrect",
            },
          ],
          "enableBurst": false,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "attacker",
      },
    ],
  },
  {
    "activePlayerId": "attacker",
    "effect": {
      "name": "InputCommand",
      "players": [
        {
          "command": [
            {
              "battery": 0,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 1,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 2,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 3,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 4,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 5,
              "type": "BATTERY_COMMAND",
            },
            {
              "type": "PILOT_SKILL_COMMAND",
            },
          ],
          "playerId": "attacker",
          "selectable": true,
        },
        {
          "nextTurnCommand": {
            "battery": 1,
            "type": "BATTERY_COMMAND",
          },
          "playerId": "defender",
          "selectable": false,
        },
      ],
    },
    "players": [
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "defender",
      },
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [
            {
              "correctBattery": 1,
              "period": {
                "type": "SpecialPeriod",
              },
              "type": "TurnStartBatteryCorrect",
            },
          ],
          "enableBurst": false,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "attacker",
      },
    ],
  },
]
`;

exports[`二人とも効果適用する場合でも正しく処理される: two-player-effective 1`] = `
[
  {
    "activePlayerId": "attacker",
    "effect": {
      "burst": {
        "recoverBattery": 5,
        "turnStartBatteryCorrect": 1,
        "type": "RecoverBattery",
      },
      "burstPlayer": "attacker",
      "name": "BurstEffect",
    },
    "players": [
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "defender",
      },
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [
            {
              "correctBattery": 1,
              "period": {
                "type": "SpecialPeriod",
              },
              "type": "TurnStartBatteryCorrect",
            },
          ],
          "enableBurst": false,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "attacker",
      },
    ],
  },
  {
    "activePlayerId": "attacker",
    "effect": {
      "invokerId": "defender",
      "name": "PilotSkillEffect",
      "skill": {
        "recoverBattery": 2,
        "type": "RecoverBatterySkill",
      },
    },
    "players": [
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": false,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "defender",
      },
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [
            {
              "correctBattery": 1,
              "period": {
                "type": "SpecialPeriod",
              },
              "type": "TurnStartBatteryCorrect",
            },
          ],
          "enableBurst": false,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "attacker",
      },
    ],
  },
  {
    "activePlayerId": "attacker",
    "effect": {
      "name": "InputCommand",
      "players": [
        {
          "command": [
            {
              "battery": 0,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 1,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 2,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 3,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 4,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 5,
              "type": "BATTERY_COMMAND",
            },
            {
              "type": "PILOT_SKILL_COMMAND",
            },
          ],
          "playerId": "attacker",
          "selectable": true,
        },
        {
          "command": [
            {
              "battery": 0,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 1,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 2,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 3,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 4,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 5,
              "type": "BATTERY_COMMAND",
            },
            {
              "type": "BURST_COMMAND",
            },
          ],
          "playerId": "defender",
          "selectable": true,
        },
      ],
    },
    "players": [
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": false,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "defender",
      },
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 5,
            "turnStartBatteryCorrect": 1,
            "type": "RecoverBattery",
          },
          "effects": [
            {
              "correctBattery": 1,
              "period": {
                "type": "SpecialPeriod",
              },
              "type": "TurnStartBatteryCorrect",
            },
          ],
          "enableBurst": false,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "attacker",
      },
    ],
  },
]
`;
